referenceqvm/qvm_stabilizer.py

```diff
@@ -0,0 +1,479 @@
+                pass
+            else:
+                invalid = True
+                break
+
+        # NOTE: all_inst is set by the subclass
+        if invalid is True and self.all_inst is False:
+            raise TypeError("Some gates used are not allowed in this QAM")
+
+        # set internal program and counter to their appropriate values
+        self.program = pyquil_program
+        self.program_counter = 0
+
+        # setup quantum and classical memory
+        q_max, c_max = self.identify_bits()
+        if c_max <= 512:  # allocate at least 512 cbits (as floor)
+            c_max = 512
+        self.num_qubits = q_max
+        self.classical_memory = np.zeros(c_max).astype(bool)
+
+    def _n_qubit_tableau(self, num_qubits):
+        """
+        Construct an empty tableau for a n-qubit system
+
+        :param Int num_qubits: number of qubits represented by the tableau
+        :return: a numpy integer array representing the tableau stabilizing
+                 the |000..000> state
+        :rtype: np.ndarray
+        """
+        tableau = np.zeros((2 * num_qubits, (2 * num_qubits) + 1),
+                                dtype=int)
+
+        # set up initial state |0>^{otimes n}
+        for ii in range(2 * self.num_qubits):
+            tableau[ii, ii] = 1
+        return tableau
+
+    def transition(self, instruction):
+        """
+        Implements a full transition, including pre/post noise hooks.
+
+        :param QuilAction instruction: instruction to be executed
+        :return: if program is halted TRUE, else FALSE
+        :rtype: bool int
+        """
+        self.pre()
+        self._transition(instruction)
+        self.post()
+
+        # return HALTED (i.e. program_counter is end of program)
+        return self.program_counter == len(self.program)
+
+    def pre(self):
+        """
+        Pre-transition hook - use for noisy evolution models. Unimplemented for now.
+        """
+        pass
+
+    def post(self):
+        """
+        Post-transition hook - use for noisy evolution models. Unimplemented for now.
+        """
+        pass
+
+    def _transition(self, instruction):
+        """
+        Implements a transition on the generator matrix representing the stabilizers
+
+        :param Gate instruction: QuilAction gate to be implemented
+        """
+        if isinstance(instruction, Measurement):
+            # mutates classical memory!  I should change this...
+            self._apply_measurement(instruction)
+            self.program_counter += 1
+
+        elif isinstance(instruction, Gate):
+            # apply Gate or DefGate
+            if instruction.name == 'H':
+                self._apply_hadamard(instruction)
+            elif instruction.name == 'S':
+                self._apply_phase(instruction)
+            elif instruction.name == 'CNOT':
+                self._apply_cnot(instruction)
+            elif instruction.name == 'I':
+                pass
+            else:
```

Comment:
what about X, Y, Z?

Reply:
Future additions!  Right now this is just CHP simulator